src/dtree.ml

```diff
@@ -0,0 +1,333 @@
+(** This module provides functions to compile rewrite rules to decision trees
+    in order to pattern match the rules efficiently.  The method is explained
+    in Luc Maranget's {i Compiling Pattern Matching to Good Decision Trees},
+    {{:10.1145/1411304.1411311}DOI}. *)
+open Terms
+open Extra
+
+(** Type of the leaves of the tree.  See {!file:terms.ml}, {!recfield:rhs}. *)
+type action = (term_env, term) Bindlib.mbinder
+
+(** Tree type.  {!const:Leaf}[(t, a)] are the leaves of the tree holding the
+    targets of the rewriting as an {!type:action}; for [t], see
+    {!recfield:switch} of {!type:node_data}.  The term option evinces--if the
+    leaf has been obtained by a switch--the term that has been matched before
+    reaching the leaf.  The {!const:Node}[n] constructor allows primarily to
+    perform a switch among subtrees recorded in {!recfield:children} of [n].
+    {!const:Fail} is a matching failure. *)
+type t = Leaf of term option * action
+       | Node of node_data
+       | Fail
+
+(** Data contained in a node of the tree.  {!recfield:switch} contains the
+    term on which the switch that gave birth to this node has been performed
+    (or none if the node has been obtained by a default matrix);
+    {!recfield:swap} indicates whether the columns of the matrix have been
+    swapped before the switch, with an int indicating the column which has
+    been swapped with the first one (or None if no swap is performed) and
+    {!recfield:children} contains the subtrees. *)
+and node_data = { switch : term option
+                ; swap : int option
+                ; children : t list}
+
+(** {b Example} Given a rewrite system for a symbol [plus] given as (in
+    [plus.dk])
+    - [plus Z (S m)     → S m]
+    - [plus n Z         → n]
+    - [plus (S n) (S m) → S (S m)]
+    A possible tree might be
```

Comment:
plus is a stranger name for this function! ;-) We should probably rename it...

Reply:
Thanks to rename plus by f, and remove the reference to plus.dk because we may rename or remove this example later.

Reply:
Fixed in edfc084.